### PR TITLE
Bro export

### PR DESCRIPTION
**Pull request by ppanero**
#### What does it do?

MISP Bro Export implemented in MISPs PHP API. Login is in the Model/ and handlers in Controller/
#### Questions
- [ ] Does it require a DB change?
- [ ] Are you using it in production?
- [ ] Does it require a change in the API (PyMISP for example)?
#### Release Type:
- [ ] Major
- [X] Minor
- [] Patch

### CODE REVIEW

File: app/Console/Command/EventShell.php [hunk 1]
```diff
@@ -263,6 +263,50 @@ public function cachenids() {
...
+	public function cachebro()
+	{
+		$broHeader = "#fields indicator\tindicator_type\tmeta.source\tmeta.url\tmeta.do_notice\tmeta.if_in\n";
+		$userId = $this->args[0];
+		$user = $this->User->getAuthUser($userId);
+		$id = $this->args[1];
+		$this->Job->id = $id;
+		$format = $this->args[2];
+		$this->Job->saveField('progress', 1);
+		//$types = array_keys($this->Attribute->typeDefinitions);
```
Comment (rotanid) [inline app/Console/Command/EventShell.php]: please remove code that isn't used.

File: app/Console/Command/EventShell.php [hunk 2]
```diff
@@ -263,6 +263,50 @@ public function cachenids() {
...
+					$filename = DS . 'misp.' . $type . '.' . $user['Organisation']['name'] . '.intel';
+				}
+				$file = new File($dir->pwd() . $filename);
+				$file->write($broHeader);
+				foreach ($final as $attribute) {
+					$file->append($attribute . "\n");
+				}
+				$file->close();
+				$zip->addFile($dir->pwd() . $filename, $type . '.intel');
+				//unlink($dir->pwd() . $filename);
```
Comment (rotanid) [inline app/Console/Command/EventShell.php]: please remove code that isn't used.

File: app/Lib/Export/NidsBroExport.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
+<?php
+
+App::uses('NidsExport', 'Export');
+
+class NidsBroExport extends NidsExport {
+
+    public function export($items, $startSid, $format = "suricata", $continue = false) {
+        // set the specific format
+        //$this->format = $format;
```
Comment (rotanid) [inline app/Lib/Export/NidsBroExport.php]: please remove code that isn't used.

File: app/Model/Job.php [hunk 4]
```diff
@@ -54,6 +54,11 @@ public function cache($type, $user, $target, $isSiteAdmin = False, $jobOrg = nul
 			$type = 'nids';
 			$extra2 = isset($user['nids_sid']) ? $user['nids_sid'] : 0;
 		}
+		if ($type === 'bro') {
+			$extra = $type;
+			$type = 'bro';
+		    $extra2 = isset($user['nids_sid']) ? $user['nids_sid'] : 0;
```
Comment (rotanid) [inline app/Model/Job.php]: wrong indention here, mixed spaces/tabs

**Comment by rotanid** [comment]
hi, i left some notes in the code.

additionally, you have wrong indention in the following files, please use TABs not spaces:
NidsBroExport.php
export_alternate.ctp
HidsBroExport.php

File: app/Console/Command/EventShell.php [hunk 5]
```diff
@@ -263,6 +263,50 @@ public function cachenids() {
...
+		$types = array('ip', 'email', 'domain', 'filename', 'filehash', 'certhash', 'software', 'url'); //Bro types
+		$typeCount = count($types);
+		$dir = new Folder(APP . DS . '/tmp/cached_exports/' . $format, true, 0750);
+
+		if ($user['Role']['perm_site_admin']) {
+			$zipname = DS . 'misp.bro.ADMIN.intel.zip';
+		} else {
+			$zipname = DS . 'misp.bro.' . $user['Organisation']['name'] . '.intel.zip';
+		}
+		$zip = new ZipArchive();
```
Comment (iglocska) [inline app/Console/Command/EventShell.php:284]: We have to watch out with this. ZipArchive won't be installed for a lot of installations, unless I am mistaken.

Comment (rotanid) [inline app/Console/Command/EventShell.php:284]: yep, either introduce it as a new dependency and fail nicely if it isn't installed...
or use command line zip (look in the code, there are some places where zip is used, please use that code as a reference, as it's tested)

**Comment by ppanero** [comment]
I've cahnged everything (identions, comments, etc.) and gonna work on the command line zip instead of ZipArchive. Sorry for the errors :/

**Comment by iglocska** [comment]
Don't be sorry, thanks a lot for working on this!

**Comment by ppanero** [comment]
I've push the changes to my branch, In EventShell is now used exec(zip...) instead of ZipArchive, I think the changes are automatically added to this PR right?

File: app/Console/Command/EventShell.php [hunk 6]
```diff
@@ -263,6 +263,53 @@ public function cachenids() {
...
+		$format = $this->args[2];
+		$this->Job->saveField('progress', 1);
+		$types = array('ip', 'email', 'domain', 'filename', 'filehash', 'certhash', 'software', 'url'); //Bro types
+		$typeCount = count($types);
+		$dir = new Folder(APP . DS . '/tmp/cached_exports/' . $format, true, 0750);
+
+		if ($user['Role']['perm_site_admin']) {
+			$zipname = DS . 'misp.bro.ADMIN.intel.zip';
+		} else {
+			$zipname = DS . 'misp.bro.' . $user['Organisation']['name'] . '.intel.zip';
```
Comment (rotanid) [inline app/Console/Command/EventShell.php:282]: i'm not sure, can an organisation contain characters, that wont work as a part of the filename? if yes, we should do some checking/escaping/replacing here?

**Comment by rotanid** [comment]
thanks for the changes. i commented on one possible issue.

also:
- your PR now removes composer.json -- please dont do that!
- your PR is based on a MISP version that's over 3 weeks old, therefore github says "This branch has conflicts that must be resolved"

note: i havent tested your PR on a installation, did only look at the code.

File: app/Console/Command/EventShell.php (hunk 6 above)
Comment (iglocska) [inline app/Console/Command/EventShell.php:282]: Definitely!

Comment (ppanero) [inline app/Console/Command/EventShell.php:282]: Yeah you'r right, anyhow that is the way it is done in every chace method in EventShell.php and I think also in AttributeController.php. So the escaping/controlling is needed in all of them.

**Comment by ppanero** [comment]
@rotanid Thanks for the comments, I dont know why it deleted composer.json, I did not touch the file. I'll check more properly next time, and concerning the branch. You are right, I am trying to update it. However since it is in one of our machines we have to be careful on updating.

File: app/Console/Command/EventShell.php (hunk 6 above)
Comment (iglocska) [inline app/Console/Command/EventShell.php:282]: The EventShell methods only create / save the files going through CakePHP's File libraries that should be safe. The problem here is that the zip portion of the code allows for arbitrary code execution via the organisation names.

**Comment by ppanero** [comment]
I did the upgrade to 2.4.51, so now checks can be done, but it override 47 files, so maybe its a good idea to cancel this PR and I perform a new one? I am not sure how to proceed, is my first time collaborating in such a big project.

**Comment by rotanid** [comment]
yep, maybe a new one is better.
currently, you also include changes of several things that dont have to do with it:
PyMISP
cakephp
app/files/taxonomies
app/files/warninglists
.gitignore  (at least parts of it)
app/Plugin/CakeResque/

maybe you did "git add -A" or "git commit -a ..." and added every local change to the repository this way...

**Comment by ppanero** [comment]
@rotanid yeah I noticed, although in fact files are not changed but overrided from what I understood. I think it is because I fetch from 2.4.51, and this PR is on 2.4... The current working release is 2.4.51 or 2.4?

**Comment by rotanid** [comment]
always use the 2.4 branch, every other one is work-in-progress stuff from @iglocska or me or other devs

**Comment by rotanid** [comment]
always use the 2.4 branch, every other one is work-in-progress stuff from @iglocska or me or other devs
@ppanero maybe you could visit us on the Gitter chat or in xmpp chat?

**Pull request closed without merging**